pay-api/src/pay_api/services/gcp_queue_publisher.py

```diff
@@ -0,0 +1,54 @@
+"""This module provides Queue type services."""
+import base64
+import json
+import uuid
+from concurrent.futures import CancelledError
+from concurrent.futures import TimeoutError  # pylint: disable=W0622
+
+from flask import current_app
+from google.auth import jwt
+from google.cloud import pubsub_v1
+from simple_cloudevent import SimpleCloudEvent, to_queue_message
+
+from .invoice import Invoice
+
+
+def publish_to_queue(payload: dict, invoice: Invoice):
+    """Publish a 'COMPLETED' invoice's info to the GCP PubSub Queue."""
+    ce = SimpleCloudEvent()
+    ce.id = payload.get('paymentToken', {}).get('id', str(uuid.uuid4()))
+    ce.source = 'sbc-pay'
+    ce.subject = invoice.business_identifier
+    ce.time = invoice.payment_date
+    ce.type = 'payment'
+    ce.data = payload
+
+    _send_to_queue(to_queue_message(ce))
+
+
+def _send_to_queue(payload: bytes):
+    """Send payload to the queue."""
+    if not ((gcp_auth_key := current_app.config.get('gcp_auth_key')) and
+            (audience := current_app.config.get('audience')) and
+            (topic_name := current_app.config.get('topic_name')) and
+            (publisher_audience := current_app.config.get('publisher_audience'))):
+
+        raise Exception('missing setup arguments')  # pylint: disable=W0719
+
+    # get authenticated publisher
+    try:
+        service_account_info = json.loads(base64.b64decode(gcp_auth_key).decode('utf-8'))
+        credentials = jwt.Credentials.from_service_account_info(
+            service_account_info, audience=audience
+        )
+        credentials_pub = credentials.with_claims(audience=publisher_audience)
+        publisher = pubsub_v1.PublisherClient(credentials=credentials_pub)
+    except Exception as error:  # noqa: B902
+        raise Exception('Unable to create a connection', error) from error  # pylint: disable=W0719
```

Comment:
Might want to create our own exceptions for these?

Reply:
When I move this to the common library, I will do that. I bumped into some changes between py3.8 and ^3.9 that caused some issues.